### ISSUE TITLE
Trouble connecting to slack

### ISSUE THREAD

**Issue by wdouglass** [lohkatsun/ws-client-egg#1]
Slack's websocket api works as follows:
you post to "https://slack.com/api/apps.connections.open" and it replies with a websocket address

When i try connecting to that websocket address, ws-connect has this error
```Error: (ws-connect) opening handshake unsuccessful: 400 Bad Request

	Call history:

	ws-client.scm:331: intarweb#headers	  
	ws-client.scm:341: intarweb#make-request	  
	ws-client.scm:341: intarweb#write-request	  
	ws-client.scm:341: intarweb#request-port	  
	ws-client.scm:352: intarweb#read-response	  
	ws-client.scm:353: intarweb#response-headers	  
	ws-client.scm:354: intarweb#response-code	  
	ws-client.scm:386: chicken.base#open-output-string	  
	ws-client.scm:386: ##sys#print	  
	ws-client.scm:387: intarweb#response-code	  
	ws-client.scm:386: ##sys#print	  
	ws-client.scm:386: ##sys#write-char-0	  
	ws-client.scm:387: intarweb#response-reason	  
	ws-client.scm:386: ##sys#print	  
	ws-client.scm:386: chicken.base#get-output-string	  
	ws-client.scm:385: ws-error	  	<--
```

Not sure if there's some feature needed by the slack api, or if i need to somehow attach my auth token to the ws request. How should i go about debugging this issue?

Thanks for any advice.

**Comment by ghost** [comment]
Is this Slack's Socket Mode API? I'm afraid I'm not familiar with it, but from a quick read through the docs just now, it doesn't seem like anything out of the usual is being asked of the ws request.

From the line numbers it seems that the error is being thrown because the server has responded to the client opening handshake with something other than a `101 Switching Protocols`. The code should be reporting the response code & reason — is anything being written to `stderr`?

**Comment by wdouglass** [comment]
yeah, the response is `Error: (ws-connect) opening handshake unsuccessful: 400 Bad Request
`. I meant to include that in the blockquote above, but it looks like the top line got cut off

I may download their python module in the morning to see if the headers are any different from the ws-client request -- i'll post what i find here

**Comment by wdouglass** [comment]
ok, i found the difference, and it looks like a behavior in `intarweb`

here's the get line that the python module emits when connecting via websocket
```
GET / HTTP/1.1
```
and here's the one that ws-client emits
```
GET wss://wss-primary.slack.com/link/?ticket=<redacted> HTTP/1.1
```

I think slack's webserver just wants the root reference in the get line, not the whole uri. I'm investigating intarweb to see if i can force this behavior somehow

**Comment by wdouglass** [comment]
ok, i got it working on my side by removing the host and port from the uri when generating the request (they're included in the request header anyway...)
I'm not sure if this functionality should be optional or is broken in some way -- i can set up a PR for this if you want

I am new to both chicken scheme and websockets, so if this is idiomatically wrong in some way please advise! thank you

the patch applies against c78efa65d7c2416919146be99cba0850c1a9b5ac. i renamed it with a .txt extension to make github happy.
[0001-remove-scheme-and-host-from-uri-when-writing-out-req.patch.txt](https://github.com/lohkatsun/ws-client-egg/files/8003566/0001-remove-scheme-and-host-from-uri-when-writing-out-req.patch.txt)

**Comment by ghost** [comment]
merged the patch for this behaviour at 92e009222a92fad2502c1aaf98d21f196033956b. Thanks again for working on this!
